Support ViPErLEED JAX plugin

alexmiame opened the pull request:
This PR collects all changes that are required for compatibility with the plugin. To make it more manageable to review, some parts that can stand on their own are split off into separate PRs. These are dependencies that need to be addressed first:
- [ ] #428  (`DISPLACEMENTS` parser)
- [x] #450 (search job class and more robust spawn vs. fork handling)
- [x] #276 (linter; *optional*)
- [ ] #500 (R-factor implementation)

michele-riva commented:
@amimre: perhaps one future-thinking option to already implement in here: rather than creating a new `SEARCH_BACKEND` parameter, we could have the new parameter be

```
BACKEND [search] = ...
```

with default `'TensErLEED'`. We could even deprecate the `TL_VERSION` parameter in favor of specifying the version in `BACKEND`.

Having it defined this way will allow us (if we ever want to) to include different backends for the calculations.

fkraushofer commented:
I'm trying this now, just submitted a job and waiting in the queue. Some feedback for as far as I got:

1. Some (at least barebones) installation instruction would be nice (maybe a dedicated page on the wiki for the JAX mode?). basically it's anyway just: download the on-the-fly-deltas, pip install; and make sure the viperleed version is this branch. correct?

2. In the docs, please add a page on `BACKEND`, and link all new parameters on the "list of parameters" pages

3. In the docs for `VLJ_CONFIG`:
> `recalc_ref_t_matrices`: Toggle whether to re-calculate the t-matrices for the reference structure or to use the t-matrices as read in from the tensor files (calculated in the [Reference calculation](https://www.viperleed.org/viperleed-jax-plugin/content/calc/sections/ref-calc.html#ref-calc)). The t-matrix calculation in viperleed-jax is uses different (more efficient and accurate) algorithms than the TensErLEED reference calculation. This leads to slightly different results for the t-matrices of the reference structure. Since the calculation of amplitude changes uses the differences between the t-matrices of the reference structure and the perturbed structure, this can lead to slightly different results for the amplitude and thus the -factor. Default: False.

It's not clear which one it `True` and which one is `False`.

> `t-leed-l_max`: Maximum value of the angular momentum quantum number l to be used in the tensor-LEED calculation. Default: Use the maximum value of [LMAX](https://www.viperleed.org/viperleed-jax-plugin/content/calc/parameters/lmax.html#lmax).

Also unclear what the "maximum value" is and how it relates to `LMAX`.

4. First attempt to execute failed: clinamen2 was missing (apparently not installed automatically)

5. Using an old DISPLACEMENTS file, I got the following error:
```
viperleed.calc.files.new_displacements.errors.DisplacementsSyntaxError: Could not parse line '<loop>'.
```
I'm not sure if it's intentional that this should raise. More critically, I had it set to `RUN = 0`, so there's no good reason that it should fail in the first place.

6. related to 5: It would be good to update the documentation for `DISPLACEMENTS` to match the new parser (also in terms of: no step needed, how does it treat multiple blocks, etc)

fkraushofer commented:
I'm not sure if this is a bug or just an incorrect warning; refcalcs seemed to run fine:
```
STARTING SECTION: REFERENCE CALCULATION
Compiling fortran files...
/gpfs/opt/sw/skylake/spack-0.19.0/opt/spack/linux-almalinux8-skylake_avx512/intel-2021.7.1/miniconda3-4.10.3-j6woa7k4wdl2m5bei4ism6tcjfp342df/lib/python3.9/multiprocessing/popen_fork.py:66: RuntimeWarning: os.fork() was called. os.fork() is incompatible with multithreaded code, and JAX is multithreaded, so this will likely lead to a deadlock.
  self.pid = os.fork()
Running reference calculations...
Reference calculations finished. Processing files...
```

amimre commented:
The bug regarding reading tensor files and the delta section should be taken care of now hopefully.

alexmiame committed:
doc: add vlj parameters to name list

alexmiame committed:
doc: clear up doc on recalc_ref_t_matrices

alexmiame committed:
doc: clear up VLJ_CONFIG t_leed_max

fkraushofer commented:
> The bug regarding reading tensor files and the delta section should be taken care of now hopefully.

Still buggy, I'm afraid:

```
Starting to interpret tensor file Tensors_003.
### ERROR ### in run_sections:run_section:236
# Error in section SEARCH
#############
### ERROR ### in run_sections:section_loop:401
# Exception during viperleed.calc execution: 
#############
Traceback (most recent call last):
  File "/home/fs71704/fkraushofer/viperleed/viperleed-jax/viperleed/src/viperleed/calc/sections/run_sections.py", line 288, in section_loop
    run_section(sec, sl, rp)
  File "/home/fs71704/fkraushofer/viperleed/viperleed-jax/viperleed/src/viperleed/calc/sections/run_sections.py", line 226, in run_section
    vlj_search(sl, rp)
  File "/home/fs71704/fkraushofer/viperleed/viperleed-jax/viperleed/src/viperleed/calc/vlj/search.py", line 103, in vlj_search
    calculator = setup_tl_calculator(
  File "/home/fs71704/fkraushofer/viperleed/on-the-fly-deltas/src/viperleed_jax/from_objects.py", line 99, in setup_tl_calculator
    tensors = read_tensor_zip(tensor_path, ref_calc_lmax, n_beams, n_energies)
  File "/home/fs71704/fkraushofer/viperleed/on-the-fly-deltas/src/viperleed_jax/files/tensors.py", line 329, in read_tensor_zip
    with zipfile.ZipFile(tensor_path, 'r') as zip_ref:
  File "/gpfs/opt/sw/skylake/spack-0.19.0/opt/spack/linux-almalinux8-skylake_avx512/intel-2021.7.1/miniconda3-4.10.3-j6woa7k4wdl2m5bei4ism6tcjfp342df/lib/python3.9/zipfile.py", line 1239, in __init__
    self.fp = io.open(file, filemode)
IsADirectoryError: [Errno 21] Is a directory: 'Tensors/Tensors_003'
```

Submitted yesterday night after pulling the most recent version of this branch; `RUN = 1 3`.

amimre commented:
> > The bug regarding reading tensor files and the delta section should be taken care of now hopefully.
> 
> Still buggy, I'm afraid:
> 
> ```
> Starting to interpret tensor file Tensors_003.
> ### ERROR ### in run_sections:run_section:236
> # Error in section SEARCH
> #############
> ### ERROR ### in run_sections:section_loop:401
> # Exception during viperleed.calc execution: 
> #############
> Traceback (most recent call last):
>   File "/home/fs71704/fkraushofer/viperleed/viperleed-jax/viperleed/src/viperleed/calc/sections/run_sections.py", line 288, in section_loop
>     run_section(sec, sl, rp)
>   File "/home/fs71704/fkraushofer/viperleed/viperleed-jax/viperleed/src/viperleed/calc/sections/run_sections.py", line 226, in run_section
>     vlj_search(sl, rp)
>   File "/home/fs71704/fkraushofer/viperleed/viperleed-jax/viperleed/src/viperleed/calc/vlj/search.py", line 103, in vlj_search
>     calculator = setup_tl_calculator(
>   File "/home/fs71704/fkraushofer/viperleed/on-the-fly-deltas/src/viperleed_jax/from_objects.py", line 99, in setup_tl_calculator
>     tensors = read_tensor_zip(tensor_path, ref_calc_lmax, n_beams, n_energies)
>   File "/home/fs71704/fkraushofer/viperleed/on-the-fly-deltas/src/viperleed_jax/files/tensors.py", line 329, in read_tensor_zip
>     with zipfile.ZipFile(tensor_path, 'r') as zip_ref:
>   File "/gpfs/opt/sw/skylake/spack-0.19.0/opt/spack/linux-almalinux8-skylake_avx512/intel-2021.7.1/miniconda3-4.10.3-j6woa7k4wdl2m5bei4ism6tcjfp342df/lib/python3.9/zipfile.py", line 1239, in __init__
>     self.fp = io.open(file, filemode)
> IsADirectoryError: [Errno 21] Is a directory: 'Tensors/Tensors_003'
> ```
> 
> Submitted yesterday night after pulling the most recent version of this branch; `RUN = 1 3`.

This also requires some commits from the viperleed-jax repo, you need to pull the most recent version there too.

alexmiame committed:
fix: NameError

alexmiame committed:
feat: show gradient benchmarks if needed

alexmiame committed:
fix: logical order

alexmiame committed:
doc,minor

alexmiame committed:
feat: R factor for log and bookkeeper

alexmiame committed:
fix: set mp start method to spawn

alexmiame committed:
fix,feat: detect, log and warn about mp start method

alexmiame committed:
Merge branch 'search-job-class' into viperleed-jax-plugin

alexmiame committed:
Merge branch 'search-job-class' into viperleed-jax-plugin

alexmiame committed:
feat: total occupation token

alexmiame committed:
doc: total occ docstring

alexmiame committed:
test: tests for total occupation token

alexmiame committed:
chore: imports

alexmiame committed:
feat: use total occ token

alexmiame committed:
Merge branch 'new-displacements-parser' into viperleed-jax-plugin

alexmiame committed:
fix: strip space and attrib

alexmiame committed:
feat,chore: rename of type token to mode

alexmiame committed:
chore: remove unused code

alexmiame committed:
feat: moved parameter space logging to tensor calculator

alexmiame committed:
feat: import

alexmiame committed:
feat: ensure VIBROCC is written to OUT

alexmiame committed:
feat: parameter space logging

alexmiame committed:
Merge branch 'ruff-linter-formatter' into viperleed-jax-plugin

alexmiame committed:
style: formatting

alexmiame committed:
style: imports

alexmiame committed:
feat: per beam R-factors

alexmiame committed:
fix: unclosed bracet

alexmiame committed:
feat: log potential shift

alexmiame committed:
fix: function name

alexmiame committed:
fix: name error

alexmiame committed:
fix: AttributeError

alexmiame committed:
fix: pass full vector

alexmiame committed:
fix: format

alexmiame committed:
fix: conversion

alexmiame committed:
fix,feat: make graphical export optional

alexmiame committed:
feat: use reference param for logged R

alexmiame committed:
style, minor

alexmiame committed:
feat: pre-optimize in parametes

alexmiame committed:
feat: preoptimize v0r default

alexmiame committed:
Merge branch '0.14.1' into viperleed-jax-plugin

alexmiame committed:
fix: wrong override of starting point

alexmiame committed:
doc: VLJ_CONFIG preoptimize_V0r

alexmiame committed:
fix: missing argument in MS Rfactor

alexmiame committed:
chore: move constant to start of file

alexmiame committed:
fix: move it back

alexmiame committed:
feat: per_beam for R2, R1 and RZJ

alexmiame committed:
fix: NaNs in gradient of ZJ R factor

michele-riva committed:
Merge branch '0.14.1' into new-displacements-parser

alexmiame committed:
chore: whitespace formatting

alexmiame committed:
doc, feat: update on the DISPLACEMENTS documentation

alexmiame committed:
Merge branch 'new-displacements-parser' of https://github.com/viperleed/viperleed into new-displacements-parser

alexmiame committed:
Apply suggestions from code review

Co-authored-by: Michele Riva <riva@iap.tuwien.ac.at>

alexmiame committed:
style: apply comments

alexmiame committed:
refactor: move displacement eps constant

alexmiame committed:
style,refactor: basic formatting

alexmiame committed:
Merge branch 'search-job-class' into new-displacements-parser

alexmiame committed:
style: more autoformatting

alexmiame committed:
refactor: more formatting

alexmiame committed:
Merge branch 'search-job-class' into new-displacements-parser

alexmiame committed:
style,refactor: more changes

alexmiame committed:
style,refactor: several more comments

alexmiame committed:
fix: missing Nones

alexmiame committed:
style: various comments by Michele

alexmiame committed:
chore,refactor: many comments by Michele

alexmiame committed:
feat: add ruff setting

alexmiame committed:
fix: loop handling

alexmiame committed:
fix,test: add more tests for loop blocks

alexmiame committed:
fix,feat: case insensitve perturbation mode

alexmiame committed:
feat,refactor: more comments

alexmiame committed:
feat: various string formatings

alexmiame committed:
fix: renamed type to mode

alexmiame committed:
refactor: backends

alexmiame committed:
feat: strings for parsed lines

alexmiame committed:
doc: fixed documentation

alexmiame committed:
fix: use eps

alexmiame committed:
fix: formatting of rhs for occ line

alexmiame committed:
fix: redefinition

alexmiame committed:
fix: check offset DOF

alexmiame committed:
fix: overly strict check

alexmiame committed:
style: no type hint

alexmiame committed:
test, fix: remove invalid test

alexmiame committed:
test,fix: mark xfail

alexmiame committed:
refactor: rename test file directory

alexmiame committed:
test,feat: test without search block header

add a simple DISPLACMENETS file that has no header

alexmiame committed:
Merge branch '0.14.1' into viperleed-jax-plugin

alexmiame committed:
Merge branch 'new-displacements-parser' into viperleed-jax-plugin

michele-riva committed:
Merge branch '0.14.1' into new-displacements-parser

alexmiame committed:
chore,style,minor: file headers and doc typos

alexmiame committed:
chore: remove redundant redefinition

alexmiame committed:
fix: remove double import

alexmiame committed:
style,minor

alexmiame committed:
chore: move files back

alexmiame committed:
doc,style,minor: file headers and type hints

alexmiame committed:
minro

alexmiame committed:
refactor: remove unneeded init

alexmiame committed:
style,minor

alexmiame committed:
fix: add super init

(though probably not necessary)

alexmiame committed:
doc,fix

alexmiame committed:
refactor: detect inconsitent OFFSETS blocks

alexmiame committed:
refactor: segments

alexmiame committed:
doc,fix

alexmiame committed:
fix: validate direction string

alexmiame committed:
feat: close check for 2 and 3 D

michele-riva commented:
@amimre: a suggestion for the documentation of the DISPLACEMENTS file: it would be good to add notes in the relevant parts stating what is knowingly not supported by the JAX backend. Things that come to mind:
- azi/r
- fractional directions
- "fixed" atomic concentrations in OCC_DELTA, as discussed in https://github.com/viperleed/viperleed/pull/428/files#r2398276829
- "combined concentrations" in OCC_DELTA
- occ CONSTRAINTS other than direct linking
- explicit "Vac" (unless implemented in #428) as per https://github.com/viperleed/viperleed/pull/428/files#r2398253441

There are also some bits that have mixed support:
- since we still use the old parser for TensErLEED, `== OFFSETS` blocks are unsupported there, and `DISPLACEMENTS` files with offsets in the `*_DELTA` sections are not supported by JAX.

Also, the BACKEND documentation should probably have a ..versionadded note.

michele-riva commented:
Other stuff that I see not to be supported at a glance in the JAX backend that would need to be documented:
- selective use of integer/fractional/all beams for optimization (only all beams supported)
- (I may be wrong:) only the first search block is considered
- MAX_TL_DISPLACEMENT
- search offsets in VIBROCC (written but not used)

alexmiame committed:
doc: class headers

alexmiame committed:
refactor init subclass

alexmiame committed:
refactor

alexmiame committed:
style,minor

alexmiame committed:
fix,minor

alexmiame committed:
Merge branch 'new-displacements-parser' into viperleed-jax-plugin

alexmiame committed:
feat,fix: import file class in init

alexmiame committed:
Merge branch 'new-displacements-parser' into viperleed-jax-plugin

alexmiame committed:
fix: broken header line logic

alexmiame committed:
Merge branch 'new-displacements-parser' into viperleed-jax-plugin

alexmiame committed:
Merge branch '0.14.1' into viperleed-jax-plugin

alexmiame committed:
feat: add Michaels new smooth R factor

alexmiame committed:
feat: kwargs for alpha/beta

alexmiame committed:
chore: remove RMS synonyms

alexmiame committed:
Merge branch '0.14.1' into viperleed-jax-plugin

alexmiame committed:
fix,refactor: don't use .type

alexmiame committed:
Merge branch 'new-displacements-parser' into viperleed-jax-plugin

alexmiame committed:
fix: mode token use

alexmiame committed:
Merge branch 'new-displacements-parser' into viperleed-jax-plugin

alexmiame committed:
fix: shift interpolated away from minimum

alexmiame committed:
feat: add new r_factor_type class

alexmiame committed:
test: tests for new r_factor_type class

alexmiame committed:
feat: use rfactor type class everywhere

alexmiame committed:
feat: new r factor type in viperleed-jax interface

fkraushofer committed:
doc: add R_S

fkraushofer committed:
doc,fix: better wording

alexmiame committed:
Merge pull request #492 from viperleed/robust_r_factor_selection

Robust R factor selection

fkraushofer committed:
feat: Fallback on R_P for refcalc

If R_S was requested, this shouldn't fail the refcalc; instead, fall back on R_P.

fkraushofer committed:
doc: add todo for R_S in error calc

alexmiame committed:
fix: wrong method name

fkraushofer committed:
Merge pull request #494 from viperleed/fallback-to-pendry

Fallback on R_P for refcalc

fkraushofer committed:
fix: R_FACTOR_TYPE needs SpecialParameter default

The previous string default would cause R-factor calculations to fail because of int('pendry')

fkraushofer committed:
fix: remove debugging prints

Since print does not necessarily end up in the right place relative to logger messages, this would print at some random point in slurm outputs. @amimre  I assume it's not currently needed?

alexmiame committed:
doc: add convergence_gens and fix ftol

alexmiame committed:
feat: add convergence_gens for CMAES

alexmiame committed:
Merge branch 'viperleed-jax-plugin' of https://github.com/viperleed/viperleed into viperleed-jax-plugin

alexmiame committed:
fix: write correct intermediate slab

fixes bug mentioned in #495

fkraushofer committed:
fix: propagate removed expbeams to top

If beams have been removed from IVBEAMS, the refcalc automatically drops corresponding beams from EXPBEAMS. In the full-dynamic optimization, this change is currently not propagated back to the "main" rp object. This causes write_fd_opt_beams_pdf to fail because the expbeams labels don't match up with the actually calculated data any more.

fkraushofer committed:
No error for requested deltas in JAX backend

Ignore request for deltas if e.g. RUN=1-3, only warn. However, *do* raise an error if error calculations are requested, since those are actually not implemented and would fail with a non-helpful error message.

fkraushofer committed:
Merge pull request #499 from viperleed/skip-requested-deltas

No error for requested deltas in JAX backend

alexmiame committed:
Merge branch 'from_viperleed_jax' into python_rfactor

alexmiame committed:
feat: use numpy and scipy

alexmiame committed:
doc: adapt header

alexmiame committed:
feat: change jax and interpax to numpy/scipy

alexmiame committed:
feat: rfactor change jax for numpy

alexmiame committed:
doc: update header

alexmiame committed:
Merge branch '0.14.2' into viperleed-jax-plugin

alexmiame committed:
Merge branch 'python_rfactor' into viperleed-jax-plugin

alexmiame committed:
fix: remove non-scipy argument check

alexmiame committed:
feat,test: adapt r factor test

alexmiame committed:
Merge branch 'python_rfactor' into viperleed-jax-plugin

michele-riva commented on src/viperleed/calc/__init__.py:
```diff
@@ -48,5 +48,14 @@
 
 LOGGER = logging.getLogger(__name__)
+
+if not mp.parent_process():  # Only in the main process
```
`multiprocessing.parent_process` was introduced in python 3.8. Here a backport to 3.7:


```python
import multiprocessing

def parent_process():
    """Return the parent process of the current one.

    This is a cross-platform replacement for
    multiprocessing.parent_process() compatible
    with Python 3.7+.

    Returns
    -------
    parent : multiprocessing.Process or None
        The parent process of the current one. None if
        this is the main process or if the parent is not
        a multiprocessing.Process.
    """
    try:
        # Python 3.8+ has it natively
        return multiprocessing.parent_process()
    except AttributeError:
        pass

    # Python 3.7 fallback
    # pylint: disable-next=protected-access  # stable
    parent_pid = multiprocessing.current_process()._parent_pid
    if parent_pid is None:
        return None

    # Try to find the Process object in the active children
    for proc in multiprocessing.active_children():
        if proc.pid == parent_pid:
            return proc

    # Parent exists but is not a multiprocessing.Process
    return None
```

alexmiame committed:
fix: numpy array for assignment

alexmiame committed:
fix: gradient stopping in rfactor

alexmiame committed:
refactor,feat: rename opt history files

alexmiame committed:
fix: store R factor to bookkeeper log

alexmiame committed:
test: test for multi digit numbers in target token

fkraushofer committed:
fix: failing regex for negative numbers

The regex replacing whitespace-separated matrices for linear operation parsing would not catch e.g. "[1 -1]" because the lookahead did not account for the possibility that the next number may begin with a sign. Fixed and added tests for such cases.

alexmiame committed:
fix: 'b' in e.g. 'Pt_sub' would be parsed as direction

alexmiame committed:
test: add test for correct target, direction separation

alexmiame committed:
fix,test: correct mode token name

fkraushofer committed:
fix: conserve renamed history.npz files in cleanup

Since these files were renamed in 84ed0f02d2b6c7d71c8dbcfd9fcc162a4c27b24d, they are currently not stored in SUPP or the history

fkraushofer committed:
Merge branch 'viperleed-jax-plugin' into debug-displacements-parser

alexmiame committed:
feat: rfactor grouping function

alexmiame committed:
test: tests for grouped R factors

alexmiame committed:
Merge pull request #504 from viperleed/debug-displacements-parser

Debug new DISPLACEMENTS parser

alexmiame committed:
feat,chore: change API to use groups

alexmiame committed:
chore: unused pendry R version

alexmiame committed:
feat,WIP: split rfactor into submodules

alexmiame committed:
feat: use dynamic numeric library

alexmiame committed:
chore: remove outdated integer V0r shift

alexmiame committed:
feat,chore: remove refactored rfactors

alexmiame committed:
fix: unused import

alexmiame committed:
chore,refactor: consistent R factor names

alexmiame committed:
feat: import R factors to init

alexmiame committed:
todo: remove TODO that was decided not to be pursued

alexmiame committed:
feat,minor: add more rfactor synonyms

alexmiame committed:
fix: file and function names

alexmiame committed:
test,fix: pendry_R name

alexmiame committed:
test,refactor: move grouping tests

alexmiame committed:
test,refactor: move rfactor tests

alexmiame committed:
doc: docstring

alexmiame committed:
fix: implement stop_gradient function

See also 068ea92af96ae83a37ace921308e0b6db115bfce

alexmiame committed:
Merge branch 'python_rfactor' into viperleed-jax-plugin

alexmiame committed:
fix: force numpy array

alexmiame committed:
Merge branch 'python_rfactor' into viperleed-jax-plugin

alexmiame committed:
chore,fix: kwargs for groups

alexmiame committed:
fix: calculation via bincount

alexmiame committed:
feat: calc integer and fractional rfactors

alexmiame committed:
Merge branch 'python_rfactor' into viperleed-jax-plugin

alexmiame committed:
fix: pass tuple to make arg hashable

alexmiame committed:
fix,feat: store and print integer and fractionals

alexmiame committed:
doc: docstring for R factor grouping

alexmiame committed:
feat: add more dynamic functions

alexmiame committed:
feat: rfactor selection function

alexmiame committed:
feat: average_beam_array

moved from viperleed-jax

alexmiame committed:
feat: determine_integer_or_fractional

alexmiame committed:
feat(!): new R-factor calculation section

alexmiame committed:
feat: compatibility with new Rfactor changes

alexmiame committed:
fix: merge conflict and imports

alexmiame committed:
Merge branch 'python_rfactor' into viperleed-jax-plugin

fkraushofer committed:
fix: R_FACTOR_SYNONYMS not internal

R_FACTOR_SYNONYMS gets imported by tensorleed_jax in tensor_calculator.set_rfactor. This currently crashes because the variable has an underscore here but not in viperleed_jax.
Alternatively, one could keep the variable internal to viperleed and replace viperleed_jax.tensor_calculator.set_rfactor entirely by viperleed.calc.lib.rfactor.select_rfactor (also avoids redundancy, but I'm not sure if the functions are always supposed to do exactly the same) - maybe something for @amimre to decide

alexmiame committed:
fix: spline coefficients shape

alexmiame committed:
fix: spline coefficients shape

alexmiame committed:
fix: call getBeamCorrespondence earlier

Since alters the beamlists in-place

alexmiame committed:
fix: call getBeamCorrespondence earlier

Since alters the beamlists in-place

alexmiame committed:
fix: correct integer-or-fractional mask size

Required size is length of expbeams, not theobeams

alexmiame committed:
fix: correct integer-or-fractional mask size

Required size is length of expbeams, not theobeams

alexmiame committed:
Merge branch 'viperleed-jax-plugin' of https://github.com/viperleed/viperleed into viperleed-jax-plugin

alexmiame committed:
Merge branch 'python_rfactor' into viperleed-jax-plugin

alexmiame committed:
Revert "fix: R_FACTOR_SYNONYMS not internal"

This reverts commit 4df5e7bd8bc0a27af4c69f8ea0e9a0f431f2b2f1.

alexmiame committed:
feat: add pendry synonym

Co-authored-by: Florian Kraushofer <66359721+fkraushofer@users.noreply.github.com>

alexmiame committed:
feat: add r_smooth synonym

Co-authored-by: Florian Kraushofer <66359721+fkraushofer@users.noreply.github.com>

alexmiame committed:
fix: apply stop_gradient

alexmiame committed:
Merge branch 'python_rfactor' into viperleed-jax-plugin

fkraushofer committed:
feat: Implement basic plotting for new r-factor

R-factor analysis file is still missing, but this restores basic plotting functionality for the new R-factor function, including when using R_s.

fkraushofer committed:
Merge branch 'python_rfactor' into viperleed-jax-plugin

fkraushofer committed:
fix: type conversion

fkraushofer committed:
minor: more detail in debug messages

fkraushofer committed:
Merge branch 'python_rfactor' into viperleed-jax-plugin

fkraushofer committed:
feat: write_Rfactorpdf_from_splines

Functionality was getting bigger, moved to a new function in iorfactor

fkraushofer committed:
feat: spline cashing, built-in theory shifts

Two optimizations for the python-based R-factor:
1. When shifting V0r, do not re-build splines at the shifted coordiantes; instead, use the existing splines and evaluate them at the shifted coordinates (the splines cover the whole energy range in any case).
2. CashedSpline: Avoid re-evaluating the derivatives and values repeatedly on the same spline (e.g. also when scanning V0r, expecially the experimental data is always the same). Instead, keep the previous evaluation in memory.

fkraushofer committed:
Merge branch 'python_rfactor' into viperleed-jax-plugin

fkraushofer committed:
fix: explicit check in CashedSpline

checking id(x) is not enough; actually compare the object

fkraushofer committed:
minor: add theo_shift in R_ms

missed in previous commit

fkraushofer committed:
minor: debug message

give some additional progress info during rfactor calculation

fkraushofer committed:
Merge branch 'python_rfactor' into viperleed-jax-plugin

fkraushofer committed:
minor: delete excess space

fkraushofer committed:
feat: accept pre-computed arrays

- change 'theo' and 'exp' nomenclature in R-factor functions to 'data_1' and 'data_2'
- add doc strings
- drop R_ms (outdated version of R_s)
- allow passing *either* splines *or* pre-computed data arrays (already on the right grid) to the R-factor functions. The latter is more JAX-friendly for the experimental data, since we would otherwise have to deal with ragged arrays, where spline interpolation cannot be cleanly vectorized.

Note that this changes the call signatures of the R-factor functions, and requires also updating the JAX plugin.

fkraushofer committed:
Merge branch 'python_rfactor' into viperleed-jax-plugin

fkraushofer committed:
fix: typo from last commit

fkraushofer committed:
fix: another issue from previous commits

fkraushofer committed:
fix: Make optimizer result actually print

The .result doesn't seem to have a repr defined, this prints empty. Explicitly print the best vector.

fkraushofer committed:
fix: search loops and repeats

- reset displacement iterator after a RUN=3 block
- actually deal with all DISPLACEMENTS blocks with a RUN=3 block

TODO: implement MAX_TL_DISPLACEMENT
Co-Authored-By: Alexandra M. Imre <90624999+amimre@users.noreply.github.com>

fkraushofer committed:
fix: only reset iterator for LoopBlock

Co-Authored-By: Alexandra M. Imre <90624999+amimre@users.noreply.github.com>

fkraushofer committed:
fix: Ignore domain DISPLACEMENTS in JAX backend

- make sure the main-PARAMETERS BACKEND carries through to the domains
- do not require DISPLACEMENTS for subdomains

fkraushofer committed:
fix: Ignore domain DISPLACEMENTS in JAX backend

- make sure the main-PARAMETERS BACKEND carries through to the domains
- do not require DISPLACEMENTS for subdomains

fkraushofer committed:
Merge pull request #517 from viperleed/debug-search-loops

Debug search loops

fkraushofer committed:
fix: load SUPERLATTICE to main params

The SUPERLATTICE from domains needs to be copied to the main rparams object before generating the BEAMLIST; otherwise, the BEAMLIST will assume (1x1) and all fractional beams will be discarded.

michele-riva committed:
fix: use work/PHAESHIFTS

Like the TensErLEED backend, it is good practice to only use files that are available in the work directory, not those from the directory where the calculation was started. Here an example why: Running a 1-3 calculation that re-creates the phaseshifts, would fail to find a PHASESHIFTS file in the home directory with a non-informative message.

My specific case: I had to recalculate the phaseshifts after adding new sites. In this case, the execution does not stop after initialization.

michele-riva committed:
Merge pull request #524 from viperleed/fix/jax/paths

fix: use work/PHASESHIFTS

fkraushofer committed:
feat: Update to access R_S from TensErLEED

TODO: update in base branch
TODO: update docs
TODO: Check TensErLEED version, default to R_P if version too old
Co-Authored-By: schmid-iap <82087012+schmid-iap@users.noreply.github.com>

fkraushofer committed:
fix: Copy BEAMLIST to subdomains

fkraushofer committed:
fix: Check TL_VERSION if R_smooth is available

alexmiame committed:
feat: add more info for dot language support

As discussed with @fkraushofer

fkraushofer committed:
Update _checksums.dat for TensErLEED v2.1.0

Add checksums for new TensErLEED version v2.1.0
